[test] Small unit test improvements, including helper to make mempool transaction

amitiuttarwar opened the pull request:
Some miscellaneous improvements that came up when working on #21061 
- The first commit is a helper to make valid mempool transactions & submit via ATMP. Introducing in this PR, using in #21061.
- The second commit is a small improvement in `miner_tests.cpp` that uses `BOOST_REQUIRE_EQUAL` to properly terminate the program instead of segfaulting in the failure scenario where the blocks do not include the expected number of transactions. 
- The third commit changes the function signature of `GetMockTime()` to return a chrono type.
- The fourth & fifth commit overload `SetMockTime` to also accept chrono type, and adds documentation to indicate that the `int64_t` function signature is deprecated.

amitiuttarwar committed:
[test] Introduce a unit test helper to create a valid mempool transaction.

amitiuttarwar committed:
[test] Throw error instead of segfaulting in failure scenario

If the miner code is faulty and does not include any transactions in a block,
the code segfaults when it tries to access block transactions. Instead, add a
check that safely aborts the process.

amitiuttarwar committed:
[doc / util] Use comments to clarify time unit for int64_t type.

amitiuttarwar commented:
the failing test seems unrelated. The failure is in `feature_assumevalid.py`, which seems hard to impact from changes that exclusively touch the unit tests, unit test framework, and some comments :)

laanwj commented on src/util/time.h:
```diff
@@ -40,9 +40,16 @@ int64_t GetTimeMicros();
+/**
+ * For testing. Set e.g. with the setmocktime rpc, or -mocktime argument
+ *
+ * @param[in] nMockTimeIn Time in seconds.
```
Probably more of a hassle, but another way to document would be to use `std::chrono` types with explicit units.

amitiuttarwar replied:
yeah, I took a quick look at changing the signature to a chrono type, but ended up taking the efficient/lazy way for now. there are a couple tweaks that would need to be made to switch it over, nothing difficult but not currently at the top of my priority list

so I thought leaving a comment was the smallest way to help :)

amitiuttarwar replied:
in the latest push, I updated the function signature of `GetMockTime`, with updated call sites. I also overloaded `SetMockTime` to take in chronos. This does mean I introduced another function that doesn't get used until #21061, so I can remove if you'd prefer.

laanwj commented:
Code review ACK e7573304b7e112b6b7f49c79c25fce36a5440209

The new function is added but not used (so not tested), I would prefer if it was, but I guess it can wait until #21061.

amitiuttarwar committed:
[util] Change GetMockTime to return chrono type instead of int

vasild commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+
+    // Add transaction to the mempool
+    TxValidationState state;
+    WITH_LOCK(cs_main, AcceptToMemoryPool(*m_node.mempool.get(), state, MakeTransactionRef(mempool_txn), nullptr, false));
```
`assert()` or `BOOST_REQUIRE()` a success here?

amitiuttarwar replied:
done

vasild commented on src/test/util/setup_common.h:
```diff
@@ -121,6 +121,11 @@ struct TestChain100Setup : public RegTestingSetup {
     CBlock CreateAndProcessBlock(const std::vector<CMutableTransaction>& txns,
                                  const CScript& scriptPubKey);
 
+    CMutableTransaction CreateValidMempoolTransaction(CTransactionRef input_transaction,
+                                                      int input_vout,
+                                                      CScript output_destination,
+                                                      CAmount output_amount = CAmount(1 * COIN));
```
The implementation assumes that the key to spend `input_transaction:input_vout` is `this->coinbaseKey`. Indeed all callers of this in #21061 pass some tx from `this->m_coinbase_txns[]`.

This is not clear from the above prototype. Maybe document this explicitly or even better - pass the spend key here, together with `input_transaction` and `input_vout`.

amitiuttarwar replied:
good point! extracted `input_signing_key` as a param to be passed in.

amitiuttarwar commented:
@vasild- thanks for the review! I've taken all your suggestions locally, but I'm currently getting a linker error when trying to `#include <boost/test/unit_test.hpp>` in `setup_common.cpp` (to use `BOOST_REQUIRE`). I'll dig into it more tomorrow, but just dropping a line here incase you (or anybody else) knows the cause off hand.

vasild commented:
@amitiuttarwar, I see no boost code is used in `setup_common.cpp`. Maybe it would be easier to use `assert()` inside `CreateValidMempoolTransaction()` instead or somehow signal a failure from that method and make the callers use `BOOST_REQUIRE()` to ensure that it succeeded.

MarcoFalke commented:
It is only allowed to use boost in the unit tests, not the fuzz tests (or other tests). setup_common is used by all test and bench libraries.

amitiuttarwar committed:
[util] Introduce a SetMockTime that takes chrono time

amitiuttarwar commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+
+    // Create an input
+    COutPoint outpoint_to_spend = COutPoint(input_transaction->GetHash(), input_vout);
+    CTxIn input(outpoint_to_spend, CScript(), CTxIn::SEQUENCE_FINAL);
```
done

amitiuttarwar commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+    AddCoins(coins_cache, *input_transaction.get(), /* check_for_overwrite */ false);
+    // - Use GetCoin to properly populate utxo_to_spend,
+    Coin utxo_to_spend;
+    coins_cache.GetCoin(outpoint_to_spend, utxo_to_spend);
```
done

amitiuttarwar commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+    CMutableTransaction mempool_txn;
+
+    // Create an input
+    COutPoint outpoint_to_spend = COutPoint(input_transaction->GetHash(), input_vout);
```
done

amitiuttarwar commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+    // - Populate a CoinsViewCache with the unspent output
+    CCoinsView coins_view;
+    CCoinsViewCache coins_cache(&coins_view);
+    AddCoins(coins_cache, *input_transaction.get(), /* check_for_overwrite */ false);
```
good catch! I updated so the height is a param that's passed in. It seems to not matter for my use cases, but since this is a helper it definitely is better to have consistency in the data structure. thanks :)

amitiuttarwar commented on src/test/util/setup_common.cpp:
```diff
@@ -231,6 +231,50 @@ CBlock TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransa
+    // - Default signature hashing type
+    int nHashType = SIGHASH_ALL;
+    std::map<int, std::string> input_errors;
+    SignTransaction(mempool_txn, &keystore, input_coins, nHashType, input_errors);
```
done

amitiuttarwar commented:
ah interesting. used `asserts` instead. thanks! 

took all the feedback. 
- added asserts to `CreateValidMempool` helper
- changed `GetMockTime` to return chrono time, updated the call sites
- introduced a `SetMockTime` that takes in chrono time, did not update call sites, but documented in the notes that the `int64_t` is deprecated and the chrono one should be used

this means that in this commit, I introduce `CreateValidMempoolTransaction` and `SetMockTime(std::chrono::seconds)`, but they are currently unused. I tested them by locally applying #21061, but ofc there is no guarantee that PR will be merged. if reviewers prefer, I can remove commit 88e62f78a7920bbeec01a05126bc8dd7b1407c4b `Introduce a SetTMockTime that takes chrono time`.

amitiuttarwar commented:
fixed a silent merge conflict since the function signature of ATMP has changed. updated #21061 to use e020cd6 to test the helper.

vasild commented on src/logging.cpp:
```diff
@@ -203,9 +203,9 @@ std::string BCLog::Logger::LogTimestampStr(const std::string& str)
-        if (mocktime) {
-            strStamped += " (mocktime: " + FormatISO8601DateTime(mocktime) + ")";
+        std::chrono::seconds mocktime = GetMockTime();
+        if (mocktime.count() == 0) {
```
Shouldn't that be `!=` instead of `==`? Or:

```suggestion
        if (mocktime > 0) {
```

amitiuttarwar replied:
🤦‍♀️ good catch.

amitiuttarwar replied:
fixed. can't compare chrono time directly to an int, so I updated to `if (mocktime > 0s)` to keep the comparison in chronos.

vasild commented on src/logging.cpp:
```diff
@@ -203,9 +203,9 @@ std::string BCLog::Logger::LogTimestampStr(const std::string& str)
-            strStamped += " (mocktime: " + FormatISO8601DateTime(mocktime) + ")";
+        std::chrono::seconds mocktime = GetMockTime();
+        if (mocktime.count() == 0) {
+            strStamped += " (mocktime: " + FormatISO8601DateTime(mocktime.count()) + ")";
```
```suggestion
            strStamped += " (mocktime: " + FormatISO8601DateTime(count_seconds(mocktime)) + ")";
```

This will prevent silent surprises if `GetMockTime()` is changed some day to return something else than `std::chrono::seconds`, e.g. `std::chrono:milliseconds`.

amitiuttarwar replied:
cool, didn't know about these helpers. updated. thanks!

amitiuttarwar commented:
updated to incorporate feedback from @vasild

vasild commented:
ACK 903bbf7f1068da0c27b99401483444f037a17840

amitiuttarwar commented:
rebased

vasild commented:
ACK 1363b6c27dbd2614fd555d148ea624ed8b95f14e

jonasschnelli commented on src/test/miner_tests.cpp:
```diff
@@ -123,6 +123,7 @@ void MinerTestingSetup::TestPackageSelection(const CChainParams& chainparams, co
     m_node.mempool->addUnchecked(entry.Fee(50000).Time(GetTime()).SpendsCoinbase(false).FromTx(tx));
 
     std::unique_ptr<CBlockTemplate> pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    BOOST_REQUIRE_EQUAL(pblocktemplate->block.vtx.size(), 4);
```
Somehow this leads to a "comparison of integers of different signs" on bitcoinbuilds.org:
https://bitcoinbuilds.org/index.php?ansilog=982c61cf-6969-4001-bebc-dc215e5d29a4.log#l1906
Maybe only happening on older boost versions?
